File: app/src/main/java/com/friday/ar/plugin/file/Manifest.java
package com.friday.ar.plugin.file;


import androidx.annotation.Nullable;

import com.friday.ar.plugin.Plugin;
import com.friday.ar.plugin.security.VerificationSecurityException;

import org.jetbrains.annotations.NotNull;

public class Manifest {
    private String pluginName;
    @Nullable
    private String author;
    private String version;
    /**
     * source file of the plugin
     */
    private PluginFile sourceFile;

    public Manifest(PluginFile sourceFile, @NotNull String pluginName, @Nullable String author, @NotNull String version) {
        this.pluginName = pluginName;
        this.author = author;
        this.version = version;
        this.sourceFile = sourceFile;
    }

    @Nullable
    public String getAuthor() {
        return author;
    }

    public String getPluginName() {
        return pluginName;
    }

    public String getVersion() {
        return version;
    }

    public void setVersion(String version) {
        this.version = version;
    }

    public Plugin toPlugin() {
        Plugin returnPlugin = new Plugin();
        returnPlugin.setName(pluginName);
        returnPlugin.setAuthorName(author);
        returnPlugin.setVersionName(version);
        returnPlugin.setPluginFile(sourceFile);
        return returnPlugin;
    }

    /**
     * Represents an exception thrown when a plugins' manifest is invalid causing security issues
     *
     * @see Manifest
     **/
    public static class ManifestSecurityException extends VerificationSecurityException {
        public ManifestSecurityException(String message) {
            super("The given manifest is not valid:" + message);
        }

        /**
         * This exception indicates that an important field is missing in the manifest and the plugin could not be loaded.
         */
        public static class MissingFieldException extends ManifestSecurityException {
            public MissingFieldException(String message) {
                super("A field is missing:" + message);
            }

            /**
             * @param context Context of the missing field, for example the containing fields' name.
             * @param message Details about the missing field.
             */
            public MissingFieldException(String context, String message) {
                super("Occurred in context '" + context + "'; A field is missing:" + message);
            }
        }
    }
}

File: app/src/main/java/com/friday/ar/plugin/file/Manifest.kt
package com.friday.ar.plugin.file


import com.friday.ar.plugin.Plugin
import com.friday.ar.plugin.security.VerificationSecurityException
import org.json.JSONObject

class Manifest(
        /**
         * * source file of the plugin
         * */
        private val sourceFile: PluginFile?, val pluginName: String, val author: String?, var version: String?) {
    companion object {
        @Throws(ManifestSecurityException.MissingFieldException::class)
        fun fromJSON(jsonObject: JSONObject): Manifest {
            if (jsonObject.isNull("meta")) throw ManifestSecurityException.MissingFieldException("meta ob object is required in manifest")
            val meta = jsonObject.getJSONObject("meta")
            return Manifest(null, meta.getString("applicationName"), meta.getString("authorName"), meta.getString("versionName"))
        }
    }
    fun toPlugin(): Plugin {
        val returnPlugin = Plugin()
        returnPlugin.name = pluginName
        returnPlugin.authorName = author
        returnPlugin.versionName = version
        returnPlugin.pluginFile = sourceFile
        return returnPlugin
    }

    /**
     * Represents an exception thrown when a plugins' manifest is invalid causing security issues
     *
     * @see Manifest
     */
    open class ManifestSecurityException(message: String) : VerificationSecurityException("The given manifest is not valid:$message") {

        /**
         * This exception indicates that an important field is missing in the manifest and the plugin could not be loaded.
         */
        class MissingFieldException(message: String) : ManifestSecurityException("A field is missing:$message")
    }
}
